test/binaryen.js/expressionrunner.js

```diff
@@ -0,0 +1,156 @@
+
+// Should be unable to evaluate a local if not explicitly specified
+runner = new binaryen.ExpressionRunner(module);
+expr = runner.runAndDispose(
+  module.i32.add(
+    module.local.get(0, binaryen.i32),
+    module.i32.const(1)
+  )
+);
+assert(expr === 0);
+
+// Should handle traps properly
+runner = new binaryen.ExpressionRunner(module);
+expr = runner.runAndDispose(
+  module.unreachable()
+);
+assert(expr === 0);
+
+// Should ignore `local.tee` side-effects if just evaluating the expression
+runner = new binaryen.ExpressionRunner(module);
+expr = runner.runAndDispose(
+  module.i32.add(
+    module.local.tee(0, module.i32.const(4), binaryen.i32),
+    module.i32.const(1)
+  )
+);
+assert(JSON.stringify(binaryen.getExpressionInfo(expr)) === '{"id":14,"type":2,"value":5}');
+
+// Should preserve any side-effects if explicitly requested
+runner = new binaryen.ExpressionRunner(module, Flags.PreserveSideeffects);
+expr = runner.runAndDispose(
+  module.i32.add(
+    module.local.tee(0, module.i32.const(4), binaryen.i32),
+    module.i32.const(1)
+  )
+);
+assert(expr === 0);
+
+// Should work with temporary values if just evaluating the expression
+runner = new binaryen.ExpressionRunner(module);
+expr = runner.runAndDispose(
+  module.i32.add(
+    module.block(null, [
+      module.local.set(0, module.i32.const(2)),
+      module.local.get(0, binaryen.i32)
+    ], binaryen.i32),
+    module.block(null, [
+      module.global.set("aGlobal", module.i32.const(4)),
+      module.global.get("aGlobal", binaryen.i32)
+    ], binaryen.i32)
+  )
+);
+assert(JSON.stringify(binaryen.getExpressionInfo(expr)) === '{"id":14,"type":2,"value":6}');
+
+// Should pick up explicitly preset values
+runner = new binaryen.ExpressionRunner(module, Flags.PreserveSideeffects);
+assert(runner.setLocalValue(0, module.i32.const(3)));
+assert(runner.setGlobalValue("aGlobal", module.i32.const(4)));
+expr = runner.runAndDispose(
+  module.i32.add(
+    module.local.get(0, binaryen.i32),
+    module.global.get("aGlobal", binaryen.i32)
+  )
+);
+assert(JSON.stringify(binaryen.getExpressionInfo(expr)) === '{"id":14,"type":2,"value":7}');
+
+// Should traverse into (simple) functions if requested
+runner = new binaryen.ExpressionRunner(module, Flags.TraverseCalls);
+module.addFunction("add", binaryen.createType([ binaryen.i32, binaryen.i32 ]), binaryen.i32, [],
+  module.block(null, [
+    module.i32.add(
+      module.local.get(0, binaryen.i32),
+      module.local.get(1, binaryen.i32)
+    )
+  ], binaryen.i32)
+);
+assert(runner.setLocalValue(0, module.i32.const(1)));
+expr = runner.runAndDispose(
+  module.i32.add(
+    module.i32.add(
+      module.local.get(0, binaryen.i32),
+      module.call("add", [
+        module.i32.const(2),
+        module.i32.const(4)
+      ], binaryen.i32)
+    ),
+    module.local.get(0, binaryen.i32)
+  )
+);
+assert(JSON.stringify(binaryen.getExpressionInfo(expr)) === '{"id":14,"type":2,"value":8}');
```

Comment:
I think it would make these tests easier to understand if the JSON were not hardcoded with the raw numbers for `id` and `type`. Would it be possible to explicitly construct the expected expressions instead?

Reply:
Used the respective constants now and added a little `assertDeepEqual` to make it more easily readable.